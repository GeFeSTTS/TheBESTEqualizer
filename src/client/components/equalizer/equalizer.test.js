import React from 'react';
import {
  shallow,
} from 'enzyme';
import Equalizer from './equalizer';

import {
  storeFactory,
} from '../../test/testUtils';

jest.mock('pizzicato', () => {
  const mockPizzicato = jest.fn();
  mockPizzicato.context = jest.fn(() => ({}));
  mockPizzicato.context.createAnalyser = jest.fn(() => 'analyser');
  mockPizzicato.Effects = jest.fn(() => {});
  const effects = ['Delay', 'PingPongDelay', 'DubDelay', 'Distortion', 'Quadrafuzz', 'Flanger',
    'Reverb', 'Tremolo', 'StereoPanner', 'Compressor', 'LowPassFilter', 'HighPassFilter', 'RingModulator',
  ];
  effects.forEach((effect, el) => {
    mockPizzicato.Effects[effect] = jest.fn(() => function Effect() {
      return {
        effect: () => (`fakeEffect${el}`),
      };
    });
  });
  return {
    __esModule: true,
    default: mockPizzicato,
  };
});

const setup = (initialState = {}) => {
  const store = storeFactory(initialState);
  const wrapper = shallow(<Equalizer store={
        store
      }
  />, { disableLifecycleMethods: true }).dive().dive();
  return wrapper;
};

describe('test equalizer', () => {
  describe('test to make snapshot', () => {
    it('should render component properly', () => {
      const initialState = {
        audioData: {
          analyser: {},
          audioContext: {},
          widthCanvas: 100,
          heightCanvas: 200,
        },
      };

      const wrapper = setup(initialState);
      expect(wrapper.instance()).toMatchSnapshot();
    });
  });

  describe('test equlizer when sound is uploaded', () => {
    const initialState = {
      audioData: {
        analyser: {},
        audioContext: {},
        widthCanvas: 100,
        heightCanvas: 200,
        sound: true,
      },
    };
    let wrapper;
    let instance;
    beforeEach(() => {
      wrapper = setup(initialState);
      instance = wrapper.instance();
    });

    it('should render `play` button', () => {
      expect(wrapper.find('Button[value="Play"]').length).toBe(1);
    });

    it('should render `pause` button', () => {
      expect(wrapper.find('Button[value="Pause"]').length).toBe(1);
    });

    it('should call function after click on `play` button', () => {
      const spyPlayMethod = jest.spyOn(instance, 'playSoundFromFile');

      instance.forceUpdate();
      wrapper.find('Button[value="Play"]').simulate('click');

      expect(spyPlayMethod).toHaveBeenCalled();
    });

    it('should call function after click on `pause` button', () => {
      const spyPauseMethod = jest.spyOn(instance, 'pauseSoundFromFile');

      instance.forceUpdate();
      wrapper.find('Button[value="Pause"]').simulate('click');

      expect(spyPauseMethod).toHaveBeenCalled();
    });

    it('should call function after click on `startStreamButton` button', () => {
      const spyPauseMethod = jest.spyOn(instance, 'startMuteStream');

      instance.forceUpdate();
      wrapper.find('Button[value="Start stream"]').simulate('click');

      expect(spyPauseMethod).toHaveBeenCalled();
    });
  });

  describe('test component methods', () => {
    const initialState = {
      audioData: {
        analyser: {
          getByteFrequencyData: jest.fn(),
        },
        audioContext: {},
        widthCanvas: 100,
        heightCanvas: 200,
        cxt: {},
        sound: {
          pause: () => (
            'fakePauseMethod'
          ),
          play: () => (
            'fakePlayMethod'
          ),
          stop: () => 'fakePauseMethod',
        },
        voice: {
          play: () => 'fakePlayMethod',
          stop: () => 'fakeStopMethod',
          pause: () => 'fakePauseMethod',
          connect: () => 'fakeConnectMethod',
        },
        startMuteState: false,
        playPauseState: false,
      },
    };

    let wrapper;
    let instance;
    beforeEach(() => {
      jest.resetAllMocks();
      wrapper = setup(initialState);
      instance = wrapper.instance();
    });

    it('will test own method `setCanvasToState` which should set in state reference on canvas element', () => {
      const paramForTest = document.createElement('canvas');
      const testedFunction = instance.setCanvasToState;

      paramForTest.getContext = jest.fn(() => 'fakeCanvas');

      testedFunction(paramForTest);
      expect(instance.state.ctx).toEqual('fakeCanvas');
    });

    it('will test own method `detectStreamSoundFromMicrophone` with resolved mediaDevices', () => {
      Object.defineProperty(window.navigator, 'mediaDevices', { value: {}, writable: true });
      Object.defineProperty(window.navigator.mediaDevices, 'getUserMedia', { value: '', writable: true });
      navigator.mediaDevices.getUserMedia = () => Promise.resolve();
      instance.createSoundStream = jest.fn();

      const testedFunction = instance.detectStreamSoundFromMicrophone;
      const spyRendEqul = jest.spyOn(instance, 'createSoundStream');

      testedFunction();
      expect(spyRendEqul).toHaveBeenCalled();
    });

    it('will test own method `detectStreamSoundFromMicrophone` with rejected `getUserMedia`', () => {
      Object.defineProperty(window.navigator, 'mediaDevices', { value: {}, writable: true });
      Object.defineProperty(window.navigator.mediaDevices, 'getUserMedia', { value: '', writable: true });
      navigator.mediaDevices.getUserMedia = () => Promise.reject();
      instance.createSoundStream = jest.fn();

      const testedFunction = instance.detectStreamSoundFromMicrophone;

      testedFunction();

      expect(() => {
        throw new Error();
      }).toThrow();
    });

    it('will test own method `detectStreamSoundFromMicrophone` with rejected mediaDevices', () => {
      Object.defineProperty(window.navigator, 'mediaDevices', { value: null, writable: true });

      const testedFunction = instance.detectStreamSoundFromMicrophone;

      expect(testedFunction).toThrow();
    });

    it('will test own method `playSoundFromFile` which should call action and function', async () => {
      wrapper.setProps({
        playPauseSoundFromFileAsProp: jest.fn(() => ('fakePlayPauseSoundFromFile')),
      });
      instance.renderEqualizer = jest.fn();
      instance.forceUpdate();

      const spyPlPasSound = jest.spyOn(instance.props, 'playPauseSoundFromFileAsProp');
      const testedFunction = instance.playSoundFromFile;
      const spyRendEqul = jest.spyOn(instance, 'renderEqualizer');

      testedFunction();
      await expect(spyPlPasSound).toHaveBeenCalled();
      await expect(spyRendEqul).toHaveBeenCalled();
    });

    it('will test own method `attachFiltersToSource` which should attech five visible effects to input source', () => {
      const fakeSourceInput = {
        howMuchEffectsAttached: 0,
        addEffect: () => {
          fakeSourceInput.howMuchEffectsAttached += 1;
        },
      };
      const testedFunction = instance.attachFiltersToSource;

      testedFunction(fakeSourceInput);
      expect(fakeSourceInput.howMuchEffectsAttached).toEqual(5);
    });

    it('will test own method `removeSoundFilters` which should attech five visible effects to input source', () => {
      const fakeProps = {
        blocksData: [{
          createEffect: 'Delay',
          isVisible: true,
        },
        {
          createEffect: 'PingPongDelay',
          isVisible: true,
        },
        {
          createEffect: 'DubDelay',
          isVisible: false,
        },
        {
          createEffect: 'Distortion',
          isVisible: false,
        },
        {
          createEffect: 'Quadrafuzz',
          isVisible: true,
        },
        {
          createEffect: 'Flanger',
          isVisible: false,
        },
        ],
        audioData: {
          ...initialState.audioData,
          sound: {
            removeEffect: (effect) => {
              fakeProps.blocksData
                .forEach((effectBlock) => {
                  if (effectBlock.createEffect === effect) {
                    effectBlock.createEffect = null;
                  }
                });
            },
            disconnect: jest.fn(),
          },
        },
      };
      wrapper.setProps({
        ...fakeProps,
      });
      instance.forceUpdate();

      const testedFunction = instance.removeSoundFilters;

      testedFunction();
      expect(fakeProps.blocksData.every((el) => {
        if (el.isVisible === true) {
          return el.createEffect == null;
        }
        return true;
      })).toBeTruthy();
      expect(fakeProps.audioData.sound).toBe(undefined);
    });

    it('will test own method `startMuteStream` which should call action and function', async () => {
      instance.renderEqualizer = jest.fn();
      wrapper.setProps({
        startMuteStreamAudioAsProp: jest.fn(() => ('fakeStartMuteStream')),
      });
      instance.forceUpdate();

      const testedFunction = instance.startMuteStream;
      const spyStMtStream = jest.spyOn(instance.props, 'startMuteStreamAudioAsProp');
      const spyRendEqul = jest.spyOn(instance, 'renderEqualizer');

      testedFunction();
      await expect(spyStMtStream).toHaveBeenCalled();
      await expect(spyRendEqul).toHaveBeenCalled();
    });

    it('will test own method `startMuteStream` which should call action and function (if statement)', async () => {
      instance.renderEqualizer = jest.fn();
      wrapper.setProps({
        audioData: {
          ...initialState.audioData,
          startMuteState: true,
        },
        startMuteStreamAudioAsProp: jest.fn(() => ('fakeStartMuteStream')),
      });
      instance.forceUpdate();

      const testedFunction = instance.startMuteStream;
      const spyStMtStream = jest.spyOn(instance.props, 'startMuteStreamAudioAsProp');
      const spyRendEqul = jest.spyOn(instance, 'renderEqualizer');

      testedFunction();
      await expect(spyStMtStream).toHaveBeenCalled();
      await expect(spyRendEqul).toHaveBeenCalled();
    });

    it('will test own method `pauseSoundFromFile` which should call action and function', async () => {
      instance.renderEqualizer = jest.fn();
      wrapper.setProps({
        playPauseSoundFromFileAsProp: jest.fn(() => 'fakePlayPauseSoundFromFile'),
      });
      instance.forceUpdate();

      const testedFunction = instance.pauseSoundFromFile;
      const spyPlPasSound = jest.spyOn(instance.props, 'playPauseSoundFromFileAsProp');
      const spyRendEqul = jest.spyOn(instance, 'renderEqualizer');

      testedFunction();
      await expect(spyPlPasSound).toHaveBeenCalled();
      await expect(spyRendEqul).toHaveBeenCalled();
    });

    it('will test own method `pauseSoundFromFile` which should call action and function (tested if statement)', async () => {
      instance.renderEqualizer = jest.fn();
      wrapper.setProps({
        audioData: {
          ...initialState.audioData,
          playPauseState: true,
        },
        playPauseSoundFromFileAsProp: jest.fn(() => 'fakePlayPauseSoundFromFile'),
      });
      instance.forceUpdate();


      const testedFunction = instance.pauseSoundFromFile;
      const spyPlPasSound = jest.spyOn(instance.props, 'playPauseSoundFromFileAsProp');
      const spyRendEqul = jest.spyOn(instance, 'renderEqualizer');

      testedFunction();
      await expect(spyPlPasSound).toHaveBeenCalled();
      await expect(spyRendEqul).toHaveBeenCalled();
    });

    it('will test own method `stopSoundFromFile` which should call functions and action', async () => {
      wrapper.setState({
        ctx: {
          canvas: {
            width: 200,
            height: 100,
          },
          clearRect: () => {},
        },
      });
      wrapper.setProps({
        audioData: {
          ...initialState.audioData,
          playPauseState: true,
        },
        playPauseSoundFromFileAsProp: jest.fn(() => 'fakePlayPauseSoundFromFile'),
      });
      instance.forceUpdate();

      const testedFunction = instance.stopSoundFromFile;
      const spyPlPasSound = jest.spyOn(instance.props, 'playPauseSoundFromFileAsProp');
      const spySoundMethod = jest.spyOn(instance.props.audioData.sound, 'stop');
      const spyCanvasMethod = jest.spyOn(instance.state.ctx, 'clearRect');

      testedFunction();
      await expect(spyPlPasSound).toHaveBeenCalled();
      expect(spySoundMethod).toHaveBeenCalled();
      expect(spyCanvasMethod).toHaveBeenCalled();
    });

    it('will test own method `stopSoundFromFile` which should call functions and not call action', async () => {
      wrapper.setState({
        ctx: {
          canvas: {
            width: 200,
            height: 100,
          },
          clearRect: () => {},
        },
      });
      wrapper.setProps({
        playPauseSoundFromFileAsProp: jest.fn(() => 'fakePlayPauseSoundFromFile'),
      });
      instance.forceUpdate();

      const testedFunction = instance.stopSoundFromFile;
      const spyPlPasSound = jest.spyOn(instance.props, 'playPauseSoundFromFileAsProp');
      const spySoundMethod = jest.spyOn(instance.props.audioData.sound, 'stop');
      const spyCanvasMethod = jest.spyOn(instance.state.ctx, 'clearRect');

      testedFunction();
      await expect(spyPlPasSound).not.toHaveBeenCalled();
      expect(spySoundMethod).toHaveBeenCalled();
      expect(spyCanvasMethod).toHaveBeenCalled();
    });

    it('will test own method `renderEqualizer` which should call function and itself by one tyme', () => {
      const spyRenderEqualizer = jest.spyOn(instance, 'renderEqualizer');
      const testedFunction = instance.renderEqualizer;
      const spyRoundedRectdMethod = jest.spyOn(instance, 'roundedRect');
      const canvasMethods = ['clearRect', 'beginPath', 'moveTo', 'lineTo', 'arcTo', 'lineTo', 'fill'];
      const mockCanvasMethod = {};

      canvasMethods.forEach((method) => {
        mockCanvasMethod[method] = jest.fn();
      });
      wrapper.setState({
        ctx: {
          canvas: {
            width: 200,
            height: 100,
          },
          ...mockCanvasMethod,
        },
      });
      wrapper.setProps({
        audioData: {
          ...initialState.audioData,
          playPauseState: true,
        },
      });
      instance.forceUpdate();

      testedFunction();
      expect(spyRoundedRectdMethod).toHaveBeenCalled();
      expect(spyRenderEqualizer).toHaveBeenCalledTimes(1);
    });
  });
});
