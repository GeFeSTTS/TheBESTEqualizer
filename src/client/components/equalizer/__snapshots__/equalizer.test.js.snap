// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test equalizer test to make snapshot should render component properly 1`] = `
Equalizer {
  "attachFiltersToSource": [Function],
  "context": Object {},
  "createSoundStream": [Function],
  "detectStreamSoundFromMicrophone": [Function],
  "pauseSoundFromFile": [Function],
  "playSoundFromFile": [Function],
  "props": Object {
    "audioData": Object {
      "analyser": Object {
        "fftSize": 128,
      },
      "audioContext": Object {},
      "heightCanvas": 200,
      "widthCanvas": 100,
    },
    "blocksData": Array [
      Object {
        "createEffect": [Function],
        "effects": Object {
          "feedback": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.01,
            "value": 0,
          },
          "mix": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.001,
            "value": 0,
          },
          "time": Object {
            "maxValue": 5,
            "minValue": 0,
            "step": 0.001,
            "value": 0,
          },
        },
        "isVisible": true,
        "name": "Delay",
      },
      Object {
        "createEffect": [Function],
        "effects": Object {
          "feedback": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.01,
            "value": 0,
          },
          "mix": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.001,
            "value": 0,
          },
          "time": Object {
            "maxValue": 5,
            "minValue": 0,
            "step": 0.001,
            "value": 0,
          },
        },
        "isVisible": false,
        "name": "Ping-Pong Delay",
      },
      Object {
        "createEffect": [Function],
        "effects": Object {
          "cutoff": Object {
            "maxValue": 4000,
            "minValue": 0,
            "step": 100,
            "value": 0,
          },
          "feedback": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.01,
            "value": 0,
          },
          "mix": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.001,
            "value": 0,
          },
          "time": Object {
            "maxValue": 5,
            "minValue": 0,
            "step": 0.001,
            "value": 0,
          },
        },
        "isVisible": false,
        "name": "Dub Delay",
      },
      Object {
        "createEffect": [Function],
        "effects": Object {
          "gain": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.01,
            "value": 0,
          },
        },
        "isVisible": false,
        "name": "Distortion",
      },
      Object {
        "createEffect": [Function],
        "effects": Object {
          "highGain": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.01,
            "value": 0,
          },
          "lowGain": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.01,
            "value": 0,
          },
          "midHighGain": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.01,
            "value": 0,
          },
          "midLowGain": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.01,
            "value": 0,
          },
          "mix": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.01,
            "value": 0,
          },
        },
        "isVisible": true,
        "name": "Quadrafuzz",
      },
      Object {
        "createEffect": [Function],
        "effects": Object {
          "depth": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.01,
            "value": 0,
          },
          "feedback": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.01,
            "value": 0,
          },
          "mix": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.01,
            "value": 0,
          },
          "speed": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.01,
            "value": 0,
          },
          "time": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.01,
            "value": 0,
          },
        },
        "isVisible": true,
        "name": "Flanger",
      },
      Object {
        "createEffect": [Function],
        "effects": Object {
          "decay": Object {
            "maxValue": 3,
            "minValue": 0,
            "step": 0.01,
            "value": 0,
          },
          "mix": Object {
            "maxValue": 3,
            "minValue": 0,
            "step": 0.01,
            "value": 0,
          },
          "time": Object {
            "maxValue": 3,
            "minValue": 0,
            "step": 0.01,
            "value": 0,
          },
        },
        "isVisible": true,
        "name": "Reverb",
      },
      Object {
        "createEffect": [Function],
        "effects": Object {
          "depth": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.01,
            "value": 0,
          },
          "mix": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.01,
            "value": 0,
          },
          "speed": Object {
            "maxValue": 20,
            "minValue": 0,
            "step": 0.01,
            "value": 0,
          },
        },
        "isVisible": false,
        "name": "Tremolo",
      },
      Object {
        "createEffect": [Function],
        "effects": Object {
          "pan": Object {
            "maxValue": 1,
            "minValue": -1,
            "step": 0.01,
            "value": 0,
          },
        },
        "isVisible": true,
        "name": "Stereo panner",
      },
      Object {
        "createEffect": [Function],
        "effects": Object {
          "attack": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.001,
            "value": 0,
          },
          "knee": Object {
            "maxValue": 40,
            "minValue": 0,
            "step": 1,
            "value": 0,
          },
          "ratio": Object {
            "maxValue": 20,
            "minValue": 1,
            "step": 1,
            "value": 1,
          },
          "release": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.001,
            "value": 0,
          },
          "threshold": Object {
            "maxValue": 0,
            "minValue": -100,
            "step": 1,
            "value": 0,
          },
        },
        "isVisible": false,
        "name": "Compressor",
      },
      Object {
        "createEffect": [Function],
        "effects": Object {
          "frequency": Object {
            "maxValue": 22050,
            "minValue": 10,
            "step": 1,
            "value": 22050,
          },
          "peak": Object {
            "maxValue": 20,
            "minValue": 0.0001,
            "step": 0.0001,
            "value": 0.0001,
          },
        },
        "isVisible": false,
        "name": "Low-Pass Filter",
      },
      Object {
        "createEffect": [Function],
        "effects": Object {
          "frequency": Object {
            "maxValue": 22050,
            "minValue": 10,
            "step": 1,
            "value": 10,
          },
          "peak": Object {
            "maxValue": 20,
            "minValue": 0.0001,
            "step": 0.0001,
            "value": 0.0001,
          },
        },
        "isVisible": false,
        "name": "High-Pass Filter",
      },
      Object {
        "createEffect": [Function],
        "effects": Object {
          "distortion": Object {
            "maxValue": 50,
            "minValue": 0.2,
            "step": 0.0001,
            "value": 0.2,
          },
          "mix": Object {
            "maxValue": 1,
            "minValue": 0,
            "step": 0.01,
            "value": 0,
          },
          "speed": Object {
            "maxValue": 2000,
            "minValue": 0,
            "step": 1,
            "value": 0,
          },
        },
        "isVisible": false,
        "name": "Ring Modulator",
      },
    ],
    "createAudioDataAsProp": [Function],
    "createStreamDataAsProp": [Function],
    "playPauseSoundFromFileAsProp": [Function],
    "startCreationAudioDataAsProp": [Function],
    "startMuteStreamAudioAsProp": [Function],
    "store": Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(Symbol.observable): [Function],
    },
  },
  "refs": Object {},
  "removeSoundFilters": [Function],
  "renderEqualizer": [Function],
  "roundedRect": [Function],
  "setCanvasToState": [Function],
  "startMuteStream": [Function],
  "state": Object {
    "analyser": Object {
      "fftSize": 128,
    },
    "audioContext": Object {},
    "ctx": null,
    "numPoints": NaN,
    "uint8Array": Uint8Array [],
  },
  "stopSoundFromFile": [Function],
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_didScheduleRenderPhaseUpdate": false,
      "_dispatcher": Object {
        "readContext": [Function],
        "useCallback": [Function],
        "useContext": [Function],
        "useDebugValue": [Function],
        "useEffect": [Function],
        "useImperativeHandle": [Function],
        "useLayoutEffect": [Function],
        "useMemo": [Function],
        "useReducer": [Function],
        "useRef": [Function],
        "useState": [Function],
      },
      "_element": <Equalizer
        audioData={
          Object {
            "analyser": Object {
              "fftSize": 128,
            },
            "audioContext": Object {},
            "heightCanvas": 200,
            "widthCanvas": 100,
          }
        }
        blocksData={
          Array [
            Object {
              "createEffect": [Function],
              "effects": Object {
                "feedback": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.01,
                  "value": 0,
                },
                "mix": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.001,
                  "value": 0,
                },
                "time": Object {
                  "maxValue": 5,
                  "minValue": 0,
                  "step": 0.001,
                  "value": 0,
                },
              },
              "isVisible": true,
              "name": "Delay",
            },
            Object {
              "createEffect": [Function],
              "effects": Object {
                "feedback": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.01,
                  "value": 0,
                },
                "mix": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.001,
                  "value": 0,
                },
                "time": Object {
                  "maxValue": 5,
                  "minValue": 0,
                  "step": 0.001,
                  "value": 0,
                },
              },
              "isVisible": false,
              "name": "Ping-Pong Delay",
            },
            Object {
              "createEffect": [Function],
              "effects": Object {
                "cutoff": Object {
                  "maxValue": 4000,
                  "minValue": 0,
                  "step": 100,
                  "value": 0,
                },
                "feedback": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.01,
                  "value": 0,
                },
                "mix": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.001,
                  "value": 0,
                },
                "time": Object {
                  "maxValue": 5,
                  "minValue": 0,
                  "step": 0.001,
                  "value": 0,
                },
              },
              "isVisible": false,
              "name": "Dub Delay",
            },
            Object {
              "createEffect": [Function],
              "effects": Object {
                "gain": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.01,
                  "value": 0,
                },
              },
              "isVisible": false,
              "name": "Distortion",
            },
            Object {
              "createEffect": [Function],
              "effects": Object {
                "highGain": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.01,
                  "value": 0,
                },
                "lowGain": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.01,
                  "value": 0,
                },
                "midHighGain": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.01,
                  "value": 0,
                },
                "midLowGain": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.01,
                  "value": 0,
                },
                "mix": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.01,
                  "value": 0,
                },
              },
              "isVisible": true,
              "name": "Quadrafuzz",
            },
            Object {
              "createEffect": [Function],
              "effects": Object {
                "depth": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.01,
                  "value": 0,
                },
                "feedback": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.01,
                  "value": 0,
                },
                "mix": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.01,
                  "value": 0,
                },
                "speed": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.01,
                  "value": 0,
                },
                "time": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.01,
                  "value": 0,
                },
              },
              "isVisible": true,
              "name": "Flanger",
            },
            Object {
              "createEffect": [Function],
              "effects": Object {
                "decay": Object {
                  "maxValue": 3,
                  "minValue": 0,
                  "step": 0.01,
                  "value": 0,
                },
                "mix": Object {
                  "maxValue": 3,
                  "minValue": 0,
                  "step": 0.01,
                  "value": 0,
                },
                "time": Object {
                  "maxValue": 3,
                  "minValue": 0,
                  "step": 0.01,
                  "value": 0,
                },
              },
              "isVisible": true,
              "name": "Reverb",
            },
            Object {
              "createEffect": [Function],
              "effects": Object {
                "depth": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.01,
                  "value": 0,
                },
                "mix": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.01,
                  "value": 0,
                },
                "speed": Object {
                  "maxValue": 20,
                  "minValue": 0,
                  "step": 0.01,
                  "value": 0,
                },
              },
              "isVisible": false,
              "name": "Tremolo",
            },
            Object {
              "createEffect": [Function],
              "effects": Object {
                "pan": Object {
                  "maxValue": 1,
                  "minValue": -1,
                  "step": 0.01,
                  "value": 0,
                },
              },
              "isVisible": true,
              "name": "Stereo panner",
            },
            Object {
              "createEffect": [Function],
              "effects": Object {
                "attack": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.001,
                  "value": 0,
                },
                "knee": Object {
                  "maxValue": 40,
                  "minValue": 0,
                  "step": 1,
                  "value": 0,
                },
                "ratio": Object {
                  "maxValue": 20,
                  "minValue": 1,
                  "step": 1,
                  "value": 1,
                },
                "release": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.001,
                  "value": 0,
                },
                "threshold": Object {
                  "maxValue": 0,
                  "minValue": -100,
                  "step": 1,
                  "value": 0,
                },
              },
              "isVisible": false,
              "name": "Compressor",
            },
            Object {
              "createEffect": [Function],
              "effects": Object {
                "frequency": Object {
                  "maxValue": 22050,
                  "minValue": 10,
                  "step": 1,
                  "value": 22050,
                },
                "peak": Object {
                  "maxValue": 20,
                  "minValue": 0.0001,
                  "step": 0.0001,
                  "value": 0.0001,
                },
              },
              "isVisible": false,
              "name": "Low-Pass Filter",
            },
            Object {
              "createEffect": [Function],
              "effects": Object {
                "frequency": Object {
                  "maxValue": 22050,
                  "minValue": 10,
                  "step": 1,
                  "value": 10,
                },
                "peak": Object {
                  "maxValue": 20,
                  "minValue": 0.0001,
                  "step": 0.0001,
                  "value": 0.0001,
                },
              },
              "isVisible": false,
              "name": "High-Pass Filter",
            },
            Object {
              "createEffect": [Function],
              "effects": Object {
                "distortion": Object {
                  "maxValue": 50,
                  "minValue": 0.2,
                  "step": 0.0001,
                  "value": 0.2,
                },
                "mix": Object {
                  "maxValue": 1,
                  "minValue": 0,
                  "step": 0.01,
                  "value": 0,
                },
                "speed": Object {
                  "maxValue": 2000,
                  "minValue": 0,
                  "step": 1,
                  "value": 0,
                },
              },
              "isVisible": false,
              "name": "Ring Modulator",
            },
          ]
        }
        createAudioDataAsProp={[Function]}
        createStreamDataAsProp={[Function]}
        playPauseSoundFromFileAsProp={[Function]}
        startCreationAudioDataAsProp={[Function]}
        startMuteStreamAudioAsProp={[Function]}
        store={
          Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(Symbol.observable): [Function],
          }
        }
      />,
      "_firstWorkInProgressHook": null,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_isReRender": false,
      "_newState": null,
      "_numberOfReRenders": 0,
      "_renderPhaseUpdates": null,
      "_rendered": <div
        className="graphicEqualizer"
      >
        <div
          style={
            Object {
              "display": "none",
            }
          }
        >
          <GraphicEqualiser
            getCanvasEl={[Function]}
            height={200}
            width={100}
          />
        </div>
        <div
          style={
            Object {
              "display": "block",
            }
          }
        >
          <Memo(ConnectFunction) />
        </div>
        <div
          className="ButtonsContainer"
        >
          <Button
            className="ButtonStyleTemplate StartStreamButton"
            icon={
              <FontAwesomeIcon
                border={false}
                className="icon"
                fixedWidth={false}
                flip={null}
                icon={
                  Object {
                    "icon": Array [
                      512,
                      512,
                      Array [],
                      "f024",
                      "M349.565 98.783C295.978 98.783 251.721 64 184.348 64c-24.955 0-47.309 4.384-68.045 12.013a55.947 55.947 0 0 0 3.586-23.562C118.117 24.015 94.806 1.206 66.338.048 34.345-1.254 8 24.296 8 56c0 19.026 9.497 35.825 24 45.945V488c0 13.255 10.745 24 24 24h16c13.255 0 24-10.745 24-24v-94.4c28.311-12.064 63.582-22.122 114.435-22.122 53.588 0 97.844 34.783 165.217 34.783 48.169 0 86.667-16.294 122.505-40.858C506.84 359.452 512 349.571 512 339.045v-243.1c0-23.393-24.269-38.87-45.485-29.016-34.338 15.948-76.454 31.854-116.95 31.854z",
                    ],
                    "iconName": "flag",
                    "prefix": "fas",
                  }
                }
                inverse={false}
                listItem={false}
                mask={null}
                pull={null}
                pulse={false}
                rotation={null}
                size={null}
                spin={false}
                symbol={false}
                title=""
                transform={null}
              />
            }
            onClick={[Function]}
            value="Start stream"
          />
          <Memo(ConnectFunction)
            handleInfoFromSound={[Function]}
          />
          <div
            style={
              Object {
                "display": "block",
              }
            }
          />
          <div
            style={
              Object {
                "display": "none",
              }
            }
          />
        </div>
        <InfoAboutTrack />
      </div>,
      "_rendering": false,
      "_updater": [Circular],
      "_workInProgressHook": null,
    },
  },
}
`;
